src/util/query-parsing.ts

```diff
@@ -6,24 +6,26 @@ export { Just, Nothing };
  * Find the (raw/urlencoded) value of a query parameter from a query string.
  * This is used because the raw parse result from `qs` automatically urldecodes
  * characters prematurely for some of our parsing purposes.
+ *
+ * query paramter is parsed according to application/x-www-form-urlencoded:
+ * https://url.spec.whatwg.org/#urlencoded-parsing
  */
 export const getQueryParamValue =
   R.curry((paramName: string, queryString: string | undefined) => {
     return Maybe(queryString).map(it =>
       it.split("&").reduce((acc: string | undefined, paramString) => {
-        const [rawKey, rawValue] = splitSingleQueryParamString(paramString);
-        return rawKey === paramName ? rawValue : acc;
+        const [parsedKey, rawValue] = parseSingleQueryParamString(paramString);
+        return parsedKey === paramName ? rawValue : acc;
       }, undefined)
     );
   });
 
-function splitSingleQueryParamString(paramString: string) {
-  const bracketEqualsPos = paramString.indexOf("]=");
-  const delimiterPos =
-    bracketEqualsPos === -1 ? paramString.indexOf("=") : bracketEqualsPos + 1;
-
-  // returning [undecoded key, undecoded value]
-  return delimiterPos === -1
-    ? [paramString, ""]
-    : [paramString.slice(0, delimiterPos), paramString.slice(delimiterPos + 1)];
+function parseSingleQueryParamString(paramString: string) {
+  const firstEqualPos = paramString.indexOf("=");
+  const [parsedKey, rawValue] = firstEqualPos === -1
+    ? [decodeURIComponent(paramString.replace("+"," ")), ""]
+    : [decodeURIComponent(paramString.slice(0, paramString.indexOf("=")).replace("+"," ")),
```

Comment:
Can we reuse `firstEqualPos` here and on the line below? Definitely a micro-optimization performance-wise, but I think it helps readability a bit too.